google/cloud/spanner/results.cc

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/results.h"
+#include "google/cloud/spanner/internal/time.h"
+#include "google/cloud/optional.h"
+#include <google/spanner/v1/result_set.pb.h>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace {
+
+optional<Timestamp> GetReadTimestamp(
+    std::unique_ptr<internal::ResultSetSource> const& source) {
+  auto metadata = source->Metadata();
+  if (metadata.has_value() && metadata->has_transaction() &&
+      metadata->transaction().has_read_timestamp()) {
+    return internal::FromProto(metadata->transaction().read_timestamp());
+  }
+  return optional<Timestamp>();
+}
+
+std::int64_t GetRowsModified(
+    std::unique_ptr<internal::ResultSetSource> const& source) {
+  return source->Stats()->row_count_exact();
+}
+
+optional<std::unordered_map<std::string, std::string>> GetExecutionStats(
+    std::unique_ptr<internal::ResultSetSource> const& source) {
+  auto stats = source->Stats();
+  if (stats && stats->has_query_stats()) {
+    std::unordered_map<std::string, std::string> execution_stats;
+    auto map = stats->query_stats().fields();
```

Comment:
`auto const&` maybe? Or maybe just `for (auto const& entry : stats->query_stats().fields())` ?

Reply:
Done.